### ISSUE TITLE
metals sbt xdg specification

### ISSUE THREAD

**Issue by tgodzik** [scalameta/metals#5796]
### Discussed in https://github.com/scalameta/metals/discussions/5775

<div type='discussions-op-text'>

<sup>Originally posted by **untainsYD** October 21, 2023</sup>
When I import an `sbt` build in `metals` (Linux `6.5.8-arch1-1`/`VSCode`), it creates a `~/.sbt` folder in the home directory, ignoring the XDG specification. Even though I have set `SBT_OPTS` in `.zprofile` as follows, it is still being ignored:
```sh
export SBT_OPTS="$SBT_OPTS \
    -Dsbt.ivy.home=$XDG_CACHE_HOME/ivy \
    -Dsbt.boot.directory=$XDG_CACHE_HOME/sbt/boot \
    -Dsbt.preloaded=$XDG_CACHE_HOME/sbt/preloaded \
    -Dsbt.global.base=$XDG_CACHE_HOME/sbt \
    -Dsbt.global.staging=$XDG_CACHE_HOME/sbt/staging \
    -Dsbt.global.zinc=$XDG_CACHE_HOME/sbt/zinc \
    -Dsbt.dependency.base=$XDG_CACHE_HOME/sbt/dependency \
    -Dsbt.repository.config=$XDG_CONFIG_HOME/sbt/repositories \
    -Dsbt.global.settings=$XDG_CONFIG_HOME/sbt/global \
    -Dsbt.global.plugins=$XDG_CONFIG_HOME/sbt/plugins \
    -Dmaven.repo.local=$XDG_CACHE_HOME/maven/repository \
    -Divy.settings.dir=$XDG_CONFIG_HOME/ivy2 \
    -Divy.home=$XDG_CACHE_HOME/ivy2 \
    -Divy.cache.dir=$XDG_CACHE_HOME/ivy2/cache"
```

I have configured the `Custom sbt launcher` to propagate environment variables, but it doesn't seem to be effective.
```sh
> ls -lah ~/.sbt/boot
Permissions Size User      Group     Date Modified Name
drwxr-xr-x     - untainsYD untainsYD 21 Oct 23:27   scala-2.12.18
.rw-r--r--     0 untainsYD untainsYD 21 Oct 23:39   sbt.boot.lock
.rw-r--r--     0 untainsYD untainsYD 21 Oct 23:27   update.log
```

How can i force metals sbt to start the server in order to my `SBT_OPTS`?</div>

**Comment by untainsYD** [comment]
If you will need an sbt file or smth like that, just mention me.

**Comment by kasiaMarek** [comment]
The issue seems  to be the bsp config generated by `sbt` (`.bsp/sbt.json`). The args for starting the server are:
```
[some-path]/java -Xms100m -Xmx100m -classpath [some-path]/sbt-launch.jar -Dsbt.script=[some-path]/sbt xsbt.boot.Boot -bsp
```
Which will work if `-Dsbt.boot.directory=...` is added explicitly but won't pick up the sbt opts. I feel like maybe `sbt` should pass those options when generating the config.

If one manually adds this option to args in `.bsp/sbt.json` it will work as expected, the problem is that `sbt` regenerates `.bsp/sbt.json` a lot, so this would have to be fixed constantly.

In metals we could insert such options when we start the sbt server, but that would work under the assumption that we know how the command for starting the server looks like, which is suboptimal, since metals should be oblivious to such things.

**Comment by tgodzik** [comment]
Shouldn't the environment variable work though? We start the process, so maybe we must make sure these variables are passed to sbt? We can suggest to restart if they changed.

**Comment by tgodzik** [comment]
It seems that it's not possible to fix this while launching via launcher, instead we should check if sbt is available on PATH and use it. For most advanced users it should be on the PATH.

Alternatively, you should be able to specify path to sbt by hand in the settings and it should work. (to be checked)

**Comment by kasiaMarek** [comment]
@untainsYD did setting sbt script in metals settings actually solved the issue for you?

**Comment by untainsYD** [comment]
> @untainsYD did setting sbt script in metals settings actually solved the issue for you?

No, it didn't, i have sbt option set, but nothing changed
![image](https://github.com/scalameta/metals/assets/46616515/1f0f7880-6451-44a9-97ea-1aaff0d24b8a)

It always recreates `~/.sbt` folder in home directory.
![image](https://github.com/scalameta/metals/assets/46616515/aedfd871-c9df-4f0d-b4e7-f44c3b215908)

**Comment by tgodzik** [comment]
Looks like maybe the env variable is not forwarded to Metals. Might be useful to start VS Code using `code` command from console. We should be forwarding the environment variable correctly otherwise :thinking:

**Comment by kasiaMarek** [comment]
I don't think it will work right now. So it seems that starting `sbt` server using`.../java ... xsbt.boot.Boot -bsp` causes `.sbt` to be created (one would have to pass `Dsbt.boot.directory=` arg to it). This command isn't generated by Metals but by `sbt` via `sbt bspConfig`, so an upstream fix seems to be needed here. Here is the sbt issue https://github.com/sbt/sbt/issues/7469.

**Comment by untainsYD** [comment]
@tgodzik @kasiaMarek sorry for bothering you, but now with `Metals version: 1.3.2+31-093989ec-SNAPSHOT` last update i have `.metals` folder generated in my home user folder within the project folder `.metals`.

![image](https://github.com/scalameta/metals/assets/46616515/70965021-c168-4c4d-ac5f-b3ce550fcd50)
![image](https://github.com/scalameta/metals/assets/46616515/469cf91d-3736-4f57-b96a-f1bce09ceb9b)

**Comment by tgodzik** [comment]
I think that would be shared indexes so that we don't reindex JDK etc. which can take a long while

**Comment by untainsYD** [comment]
> I think that would be shared indexes so that we don't reindex JDK etc. which can take a long while

i got it, but having it on `XDG_PATH` would be very nice.

**Comment by untainsYD** [comment]
@kasiaMarek, is setting `-sbt-dir`, `-sbt-boot` in `/etc/sbt/subtopics` or `~/.config/sbt/subtopics` the same as setting SBT_OPTS, and will neither help?
